blockchain/consensus/deleg_cns/src/validation.rs

```diff
@@ -92,10 +92,20 @@ fn validate_miner<DB>(
 where
     DB: BlockStore + Send + Sync + 'static,
 {
+    use DelegatedConsensusError::*;
     let miner_addr = header.miner_address();
+
+    // Workaround for the bug where Forest strips the network type from the Address
```

Comment:
Do we have an issue for this bug in the forest repo?

Reply:
There were some issues related to this:
* https://github.com/ChainSafe/forest/issues/610
* https://github.com/ChainSafe/forest/pull/611
* https://github.com/ChainSafe/forest/pull/233

As I understand the problem is that the protocol doesn't serialise the network to bytes, so once we convert to `ID` the information is lost, and to be consistent with that, the string formatting is also lossy. It probably wasn't meant to appear on its own, I'm not sure if its use in JSON config and the wallet was part of the design. 

I don't think there is an issue to recognise the friction it causes, but it's possible it just wasn't revealed in my quick search.

So the issue is that 

```
let a0 = Address::from_str("t01000").unwrap();
assert_eq!(a0.network, Network::Testnet);
if let &Payload::ID(id) = a0.payload() {
    let a1 = Address::from_id(id);
    assert_eq!(a1.network, Network::Mainnet);
}
```
This is basically what's happening in [StateManager::lookup_id](https://github.com/ChainSafe/forest/blob/v0.3.0/blockchain/state_manager/src/lib.rs#L1268-L1273). 

As I reported on Slack earlier, this also affects methods like `Address::new_bls` and `Address::new_secp256k1`, where the default `Network::Mainnet` is attached, which the [wallet_helpers::new_address](https://github.com/ChainSafe/forest/blob/v0.3.0/key_management/src/wallet_helpers.rs#L29) doesn't try to alter with `set_network` because it's not part of the [WalletImportParams](https://github.com/ChainSafe/forest/blob/v0.3.0/node/rpc/src/wallet_api.rs#L110) in the first place. As I found out later it might actually be cancelled by the fact that the state tree produced by Lotus also has `f` addresses, even though in the `genesis.json` file they were `t` addresses :shrug: 

I generally found working with the `Address` confusing, seeing some code that turns it from one format to another format and clearly expecting that the `Payload` will be of a certain variant, but without the type system indicating this, so I opened some PRs in the actor code of a fork we work on for Hierarchical Consensus to improve the situation by adding a `TAddress` which tells exactly what it can be:
* https://github.com/adlrocha/builtin-actors/pull/15
* https://github.com/adlrocha/builtin-actors/pull/16

This is similar to my [earlier PR](https://github.com/filecoin-project/builtin-actors/issues/489) about typed CIDs which now live [here](https://github.com/adlrocha/builtin-actors/tree/master/primitives/src/tcid). 

Let me know if you find interest in adding these to Forest, I'm not sure what would be the best place for them, but I would not live without them.